### ISSUE TITLE
duplicated middleware upcast functions

### ISSUE THREAD

**Issue by mattsse** [prestwich/ethers-middleware-experimental#3]
Currently there are following trait functions 

```rust
Middleware::as_base_middleware();
Middleware::as_geth_middleware();
Middleware::as_parity_middleware();


GethMiddleware::as_base_middleware();

ParityMiddleware::as_base_middleware();


<NetWork>Middleware::as_base_middleware();
<NetWork>Middleware::as_geth_middleware();
<NetWork>Middleware::as_parity_middleware();
```

Now, since `trait EthereumMiddleware: BaseMiddleware<Ethereum> + GethMiddleware<Ethereum> + ParityMiddleware<Ethereum> + Middleware<Ethereum>`

the `Ethereum` middleware struct now has 4! `as_base_middleware()` functions, and 2 `as_geth_middleware` `as_parity_middleware`, which may not be obvious to the user.

In itself, this may not be a problem as long as only `networks::ethereum::Middleware` is imported, since rust then clearly knows which function to call.

However, the `EthereumMiddleware` is implemented in a macro, which makes it harder for IDEs to resolve the trait properly:

![image](https://user-images.githubusercontent.com/19890894/146946060-6a9f9b9b-a84e-4583-b3f9-fdf1caf5c4df.png)

Notice that intellij detects 3

```
Middleware::as_base_middleware();
GethMiddleware::as_base_middleware();
ParityMiddleware::as_base_middleware();
```

but fails to resolve the correct one `EthereumMiddleware::as_base_middleware`

This is the case for basically every function that is duplicated in `EthereumMiddleware`

instead of invoking the macro, implementing this by hand for `Ethereum` will definitely help. perhaps this is also only limited to working within the same repo and will be a non-issue if pulled in as a dependency. But I wonder if there's a way to reduce these duplicate functions properly?

**Comment by prestwich** [comment]
Yeah, so this problem actually exists for _all_ methods on these traits, as they're all overloaded by the `EthereumMiddleware`. It's also annoying for `send_transaction` and `send_raw_transaction` which are duplicated across `BaseMiddleware` (returns `H256`) and `Middleware` (returns `PendingTransaction`.

> perhaps this is also only limited to working within the same repo and will be a non-issue if pulled in as a dependency. But I wonder if there's a way to reduce these duplicate functions properly?

that's my belief. As long as only `EthereumMiddleware` is in scope, we have no issues. It's only when we have multiple middleware traits in scope (e.g. only for implementors) that we run into issues here

You can see a usage example in the subscriptions tests [here](https://github.com/prestwich/ethers-middleware-experimental/blob/main/src/subscriptions/mod.rs#L135). As long as only `EthereumMiddleware` is in scope, there's no need to disambiguate any method in the code. 

So it seems like the real problem here is making it as convenient to intellij and rust-analyzer as possible.

> However, the `EthereumMiddleware` is implemented in a macro, which makes it harder for IDEs to resolve the trait properly:

idle thought, can we run this macro expansion in a `build.rs`?

**Comment by prestwich** [comment]
A further idle thought: the duplicated upcasts can be solved by disambiguating `geth_as_base` from `parity_as_base`
